pkg/webhook/cvc.go

```diff
@@ -0,0 +1,199 @@
+		return response
+	}
+	err = validateCVCSpecChanges(cvcOldObj, &cvcNewObj)
+	if err != nil {
+		klog.Errorf("invalid cvc changes: %s error: %s", cvcOldObj.Name, err.Error())
+		response = BuildForAPIObject(response).UnSetAllowed().WithResultAsFailure(err, http.StatusBadRequest).AR
+		return response
+	}
+	return response
+}
+
+func validateCVCSpecChanges(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	validateFuncList := []validateFunc{validateReplicaCount,
+		validatePoolListChanges,
+		validateReplicaScaling,
+		validateStatusPoolList}
+	for _, f := range validateFuncList {
+		err := f(cvcOldObj, cvcNewObj)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Below validations should be done only with new CVC object
+	err := validatePoolNames(cvcNewObj)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// TODO: isScalingInProgress(cvcObj *apis.CStorVolumeClaim) signature need to be
+// updated to cvcObj.IsScaleingInProgress()
+func isScalingInProgress(cvcObj *apis.CStorVolumeClaim) bool {
+	return len(cvcObj.Spec.Policy.ReplicaPoolInfo) != len(cvcObj.Status.PoolInfo)
+}
+
+// validateReplicaCount returns error if user modified the replica count after
+// provisioning the volume else return nil
+func validateReplicaCount(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	if cvcOldObj.Spec.ReplicaCount != cvcNewObj.Spec.ReplicaCount {
+		return errors.Errorf(
+			"cvc %s replicaCount got modified from %d to %d",
+			cvcOldObj.Name,
+			cvcOldObj.Spec.ReplicaCount,
+			cvcNewObj.Spec.ReplicaCount,
+		)
+	}
+	return nil
+}
+
+// validatePoolListChanges returns error if user modified only the pool names
+func validatePoolListChanges(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	oldDesiredPoolNames := cvc.GetDesiredReplicaPoolNames(cvcOldObj)
+	newDesiredPoolNames := cvc.GetDesiredReplicaPoolNames(cvcNewObj)
+	modifiedPoolNames := util.ListDiff(oldDesiredPoolNames, newDesiredPoolNames)
+	if len(newDesiredPoolNames) >= len(oldDesiredPoolNames) {
+		// If no.of pools on new spec >= no.of pools in old spec(scaleup as well
+		// as migration then there all the pools in old spec must present in new
+		// spec)
+		if len(modifiedPoolNames) > 0 {
+			return errors.Errorf(
+				"volume replica migration directly by modifying pool names %v is not yet supported",
+				modifiedPoolNames,
+			)
+		}
+	} else {
+		// If no.of pools in new spec < no.of pools in old spec(scale down
+		// volume replica case) then there should at most one change in
+		// oldSpec.PoolInfo - newSpec.PoolInfo
+		if len(modifiedPoolNames) > 1 {
+			return errors.Errorf(
+				"Can't perform more than one replica scale down requested scale down count %d",
+				len(modifiedPoolNames),
+			)
+		}
+	}
+	return nil
+}
+
+// validateReplicaScaling returns error if user updated pool list when scaling is
+// already in progress
+func validateReplicaScaling(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
```

Comment:
please upated the comment with info related to scale up cases, like scale up for more then 1 replica  can be done simultaneously but not for the new scale request if scale operation is going on

Reply:
updated